cpu/nrf52/periph/pdm.c

```diff
@@ -0,0 +1,167 @@
+int16_t _pdm_buf[PDM_BUF_SIZE * 2] = { 0 };
+static uint8_t _pdm_current_buf = 0;
+static pdm_isr_ctx_t isr_ctx;
+
+int pdm_init(pdm_mode_t mode, pdm_sample_rate_t rate, int8_t gain,
+              pdm_data_cb_t cb, void *arg)
+{
+    if (NRF_CLOCK->EVENTS_HFCLKSTARTED == 0) {
+        NRF_CLOCK->TASKS_HFCLKSTART = 1;
+        while (NRF_CLOCK->EVENTS_HFCLKSTARTED == 0) {}
+    }
+
+    /* Configure sampling rate */
+    switch (rate) {
+        case PDM_SAMPLE_RATE_16KHZ:
+#ifdef CPU_MODEL_NRF52840XXAA
+            NRF_PDM->RATIO = ((PDM_RATIO_RATIO_Ratio80 << PDM_RATIO_RATIO_Pos) & PDM_RATIO_RATIO_Msk);
+            NRF_PDM->PDMCLKCTRL = PDM_PDMCLKCTRL_FREQ_1280K;
+#else
+            NRF_PDM->PDMCLKCTRL = PDM_PDMCLKCTRL_FREQ_Default; /* 1.033MHz */
+#endif
+            break;
+        case PDM_SAMPLE_RATE_20KHZ:
+            NRF_PDM->PDMCLKCTRL = 0x0A000000; /* CLK= 1.280 MHz (32 MHz / 25) => rate= 20000 Hz */
+            break;
+        case PDM_SAMPLE_RATE_41KHZ:
+            NRF_PDM->PDMCLKCTRL = 0x15000000; /* CLK= 2.667 MHz (32 MHz / 12) => rate= 41667 Hz */
+            break;
+        case PDM_SAMPLE_RATE_50KHZ:
+            NRF_PDM->PDMCLKCTRL = 0x19000000; /* CLK= 3.200 MHz (32 MHz / 10) => rate= 50000 Hz */
+            break;
+        case PDM_SAMPLE_RATE_60KHZ:
+            NRF_PDM->PDMCLKCTRL = 0x20000000; /* CLK= 4.000 MHz (32 MHz / 8) => rate= 60000 Hz */
+            break;
+        default:
+            DEBUG("[pdm] init: sample rate not supported\n");
+            return -ENOTSUP;
+    }
+
+    /* Configure mode (Mono or Stereo) */
+    switch (mode) {
+        case PDM_MODE_MONO:
+            NRF_PDM->MODE = PDM_MODE_OPERATION_Mono;
+            break;
+        case PDM_MODE_STEREO:
+            NRF_PDM->MODE = PDM_MODE_OPERATION_Stereo;
+            break;
+        default:
+            DEBUG("[pdm] init: mode not supported\n");
+            return -ENOTSUP;
+    }
+
+    /* Configure gain */
+    if (gain > PDM_GAIN_MAX) {
+        gain = PDM_GAIN_MAX;
+    }
+
+    if (gain < PDM_GAIN_MIN) {
+        gain = PDM_GAIN_MIN;
+    }
+
+    NRF_PDM->GAINR = (gain << 1) + 40;
+    NRF_PDM->GAINL = (gain << 1) + 40;
+
+    /* Configure CLK and DIN pins */
+    gpio_init(pdm_config.clk_pin, GPIO_OUT);
+    gpio_clear(pdm_config.clk_pin);
+    gpio_init(pdm_config.din_pin, GPIO_IN);
+
+    NRF_PDM->PSEL.CLK = pdm_config.clk_pin;
+    NRF_PDM->PSEL.DIN = pdm_config.din_pin;
+
+    /* clear pending events */
+    NRF_PDM->EVENTS_STARTED = 0;
+    NRF_PDM->EVENTS_STOPPED = 0;
+    NRF_PDM->EVENTS_END = 0;
+
+    /* Enable end/started/stopped events */
+    NRF_PDM->INTENSET = ((PDM_INTEN_END_Enabled << PDM_INTEN_END_Pos) |
+                        (PDM_INTEN_STARTED_Enabled << PDM_INTEN_STARTED_Pos) |
+                        (PDM_INTEN_STOPPED_Enabled << PDM_INTEN_STOPPED_Pos));
+
+    /* Configure internal RAM buffer size, divide by 2 for stereo mode */
+    NRF_PDM->SAMPLE.MAXCNT = (PDM_BUF_SIZE >> mode);
```

Comment:
```suggestion
    NRF_PDM->SAMPLE.MAXCNT = (PDM_BUF_SIZE);
```
According to the datasheet and also logically the SAMPLE.MAXCNT shouldn't change if we run the pdm in stereo mode, because we still want to fill the double buffer _pdm_buff and not only half of it.

The only thing that changes is that in stereo mode "the buffer in RAM can contain half the stereo sampling time as compared to the mono sampling time".